Improvements to the neural dual solver

bamos opened the pull request:
Hi @marcocuturi @bunnech @michalk8, here is a WIP PR with the updates we've discussed to the neural dual solver. It improves the example from:

![image](https://user-images.githubusercontent.com/707462/209060205-ab36d7c5-3edb-4a48-b864-45b466db9772.png)

to this:

<img width="374" alt="image" src="https://user-images.githubusercontent.com/707462/209724655-b9e5480b-98a8-4de7-854b-7b92a36eb5ed.png">


[My paper](https://arxiv.org/abs/2210.12153) has some more context on these updates adding some more options for fine-tuning and amortizing the conjugate approximation. I've marked this as a WIP because I'm still debugging the inverse map in the example, but otherwise, the rest of the code is ready for an initial review as it would be good to get your thoughts on the core updates and API/doc changes. I can also call and chat a little more about these soon if that would be helpful. Here are quick summaries of my updates to the code and notebook.

## Updates to the core neural solver
1. swap the order of f and g potentials (\cc https://github.com/ott-jax/ott/issues/182#issuecomment-1344895638)
2. add an option to add a conjugate solver that fine-tunes g's prediction when updating f, along with a default solver using JaxOpt's LBFGS optimizer
3. add amortization modes to learn g: regression and objective (Makkuva et al.)
4. an option to update both potentials with the same batch (in parallel) in addition to the separate inner loop Makkuva et al. uses
5. the option to model the *gradient* of the g potential with a neural network and a simple MLP implementing this instead of the ICNN from Makkuva et al.
6. a callback so the training progress can be monitored (debugging the notebook/code was very difficult without this)

## Updates to the notebook/example
(I'm still tweaking these around a little as it's still a little suboptimal)

1. Update f ICNN size from [64 ,64, 64, 64] to [128, 128]
  and modified the architecture
3. Use an MLP to model the gradient of g
4. Use the default JaxOpt LBFGS conjugate solver and
   update g to regress onto it
5. Increase batch size from 1000 to 10000
6. Sample the data in batches rather than sequentially
7. Tweak Adam (use weight decay and a cosine schedule)

I've also tried running the code with an approach closer to the older implementation, i.e., the approach from Makkuva et al. that models g with an ICNN and uses 10 inner objective-based updates, but can't get it to work as nicely as the version with conjugate fine-tuning+regression.

## Remaining discussion and TODOs
+ [x] Decide on best format for the arguments. E.g., should the conjugate solver be a function and the amortization mode a string?
+ [x] Decide on the best defaults to use (a lightweight LBFGS conjugate solver for fine-tuning that regresses g onto it, or something closer to the approach in Makkuva et al.?)
+ [x] Improve the documentation and notebook to clarify these options. 
  + Add/enable docs for the conjugate solver and any other significantly new code
+ [x] Update other places using the Neural OT code (like the initialization schemes notebook)
+ [x] Think about any tests that would be useful to add for these updates
+ [x] Fix the inverse map in the notebook
+ [x] Decide on the default ICNN architecture
+ [x] Add ability to save/load the parameters and ability to re-initialize the solves with them

marcocuturi commented on src/ott/problems/linear/potentials.py:
```diff
@@ -71,7 +71,7 @@ def transport(self, vec: jnp.ndarray, forward: bool = True) -> jnp.ndarray:
     vec = jnp.atleast_2d(vec)
     if self._corr and isinstance(self.cost_fn, costs.SqEuclidean):
-      return self._grad_g(vec) if forward else self._grad_f(vec)
+      return self._grad_f(vec) if forward else self._grad_g(vec)
```
thanks!

marcocuturi commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -12,49 +12,78 @@
+)
 
 
 class NeuralDualSolver:
```
Maybe we should start using a clearer naming convention and drop the `Solver`, since it is not used in for other solvers (e.g. `Sinkhorn` or `GromovWasserstein`). We might also want to specify this is a 2Wasserstein problem. So maybe `2WassersteinDual` ? or  `NeuralDual2Was`? or any other idea : )

bamos replied:
It could also be clean to let`NeuralDual` be the Euclidean Wasserstein-2 one since it's the most commonly used one, and then give neural OT variations more specialized names.

Slightly related: there's also the question of whether "solver" should remain in the names for the conjugate solver. It may be helpful to keep since it's not always clear the conjugate is computed with an explicit solver, but we could also consider renaming to something like `ConjugateLBFGS`

marcocuturi replied:
Maybe `W2NeuralDual` then? (we can expect someone will implement `W1` at some point)

I also think `Fenchel` or `Legendre` is a bit more explicit than `Conjugate` in convex analysis literature. Maybe introduce `Fenchel` and then use the solver by default being LBFGS? 

Another question I am wondering is, wouldnt it be a bit more "JAX functional" to have a function transform taking a function `f` in and returning another function `tilde_f`?

bamos replied:
I updated `ConjugateSolverLBFGS->FenchelConjugateLBFGS` in https://github.com/ott-jax/ott/pull/219/commits/0681a034ee797d6be4dc13213133f6bb65e6b7d1 but kept the argument name to `W2NeuralDual` as `conjugate_solver` for now. Do you think we should update this name too?

> Another question I am wondering is, wouldnt it be a bit more "JAX functional" to have a function transform taking a function f in and returning another function tilde_f?

One way this could be done is by making the solver a module that takes the function as the input and then updating `solve` to `__call__`. I originally didn't do it this way since I didn't want to construct a new function/object every time to compute the conjugate, but maybe it's not so bad and isn't much overhead

marcocuturi commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -12,49 +12,78 @@
 
 
 class NeuralDualSolver:
   r"""Solver of the ICNN-based Kantorovich dual.
```
ICNN should be dropped here if you want to consider a general MLP

marcocuturi commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -220,65 +341,75 @@ def train_neuraldual(
+      """Loss function for both potentials."""
       # get two distributions
-      source, target = batch["source"], batch["target"]
+      X, Y = batch["source"], batch["target"]
```
we usually avoid capital letters for variable names.

bamos replied:
reverted

marcocuturi reviewed (commented):
Hi Brandon!!! Thanks for the holidays present :) This looks really nice! Just a very short review to start with some comments, will go deeper after holidays are over.

marcocuturi commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -108,55 +146,138 @@ def setup(
+            params=self._clip_weights_icnn(self.state_f.params)
+        )
+
+      # log to wandb
```
is this wandb specific?

bamos replied:
I didn't add this part, but it doesn't seem like it. Removed in https://github.com/ott-jax/ott/pull/219/commits/827abe4747d74e2580c89342c20374abbd10c0be

marcocuturi commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -108,55 +146,138 @@ def setup(
+      train_batch["source"] = jnp.array(next(trainloader_source))
+      train_batch["target"] = jnp.array(next(trainloader_target))
+
+      self.state_f, self.state_g, loss, loss_f, loss_g, w_dist = \
```
let's avoid the line split :)

bamos replied:
updated these

marcocuturi commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -108,55 +146,138 @@ def setup(
+        valid_batch["source"] = jnp.array(next(validloader_source))
+        valid_batch["target"] = jnp.array(next(validloader_target))
+
+        valid_loss_f, valid_loss_g, valid_w_dist = \
```
same for line split

marcocuturi reviewed (commented):
thanks Brandon, a few more coments. 

Should we discuss more specifically the items you have outlined in the PR?

bamos commented on src/ott/solvers/nn/icnn.py:
```diff
@@ -182,15 +174,11 @@ def _compute_identity_map(input_dim: int) -> Tuple[jnp.ndarray, jnp.ndarray]:
+    for i in range(self.num_hidden):
+      z = jnp.add(self.w_zs[i](z), self.w_xs[i + 1](x))
+      z = self.act_fn(z)
+    z += self.pos_def_potential(x)
```
Yes, this change significantly helps: reverting this and re-running the notebook will result in a significantly worse result. I'm not sure what the best intuition/reason is for this and it could just be related to the initialization, but I think giving this potential's scalar value to the convex path at the beginning rather than the input makes it hard for the convex path to learn the right transformation of the input since it's never able to directly access it.

It could work to concatenate the quadratic potential's value to `z` in L177 so the following convex layers would have access to both. It may be adding too many parameters but we could also parameterize whether all of these are added (1. if the quadratic potential is used at the beginning or end of the network, or both, and 2. if the convex path initially accesses `x`)

marcocuturi replied:
> the convex path to learn the right transformation of the input since it's never able to directly access it.

not sure I follow what you mean by "never" here? Do you mean some kind of vanishing gradient problem that prevents the linear part in the gradient (hence quadratic in the potential) to be meaningfully updated? If that's the case I think the intuition kind of makes sense! it's no longer a "proper" ICNN as you had defined in your paper, but this works :)

bamos replied:
I wasn't thinking about the vanishing gradient w.r.t. the quadratic potential (although that makes sense too), but was referring to how the previous version with the quadratic potential at the beginning transforms the input into a single scalar-valued potential. It seems like it could create a bottleneck because the following layers in the input-convex path need to base the input-convex path on transformations of the initial potential's scalar value rather than of the vector-valued input.

bamos commented on src/ott/solvers/nn/mlp.py:
```diff
@@ -0,0 +1,61 @@
+__all__ = ["PotentialGradientMLP"]
+
+
+class PotentialGradientMLP(nn.Module):
```
Hmm, what about just `MLP` then?

bamos commented on src/ott/solvers/nn/mlp.py:
```diff
@@ -0,0 +1,61 @@
+  dim_hidden: Sequence[int]
+
+  @property
+  def provides_gradient(self):
```
Updated in https://github.com/ott-jax/ott/pull/219/commits/aac7386afbccdd465f7a4cb7455fcca4a29eb00f

bamos commented:
(I'll fix that test soon too)

michalk8 commented on docs/references.bib:
```diff
@@ -665,3 +665,10 @@ @InProceedings{sejourne:22
   url = {https://proceedings.mlr.press/v151/sejourne22a/sejourne22a.pdf},
 }
+
+@article{amos:22a,
```
Can you please use 
```bibtex
@misc{amos:22a,
  doi = {10.48550/ARXIV.2210.12153},
  url = {https://arxiv.org/abs/2210.12153},
  author = {Amos, Brandon},
  keywords = {Machine Learning (cs.LG), Artificial Intelligence (cs.AI), Machine Learning (stat.ML), FOS: Computer and information sciences, FOS: Computer and information sciences},
  title = {On amortizing convex conjugates for optimal transport},
  publisher = {arXiv},
  year = {2022},
  copyright = {Creative Commons Attribution 4.0 International}
}
```
as it contains also the URLs?

michalk8 commented on src/ott/solvers/nn/icnn.py:
```diff
@@ -104,25 +98,23 @@ def setup(self) -> None:
+    self.pos_def_potential = PosDefPotentials(
+        self.dim_data,
+        num_potentials=1,
+        kernel_init=lambda *args, **kwargs: factor,
```
Would use `*_, **__` to indicate these are throw-away arguments.

michalk8 commented on src/ott/solvers/nn/mlp.py:
```diff
@@ -0,0 +1,61 @@
+
+    if single:
+      z = jnp.squeeze(z, 0)
+    return z
```
Would add it as a one-liner, e.g., `return z.squeeze(0) if single else z`

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -12,49 +12,81 @@
+  :math:`\nabla f` transports source to target cells.
+  This potential is learned by optimizing the dual
+  form associated with the negative inner product cost
+  :math:`\argsup_{\theta}\; -\E_{x\sim\alpha}[f_\theta(x)] - \E_{y\sim\beta}[f^\star_\theta(y)]`
```
Too long line, please split. Also consider non-inline math as:
```rst
.. math::
 ...
```

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -64,14 +96,17 @@ def __init__(
       pos_weights: bool = True,
       beta: float = 1.0,
+      conjugate_solver: Optional[Callable] = default_conjugate_solver,
+      amortization_loss: str = 'regression',
```
Should be `Literal['objective', 'regression']`.

bamos commented on pyproject.toml:
```diff
@@ -17,6 +17,7 @@ dependencies = [
     "jax>=0.1.67",
     "jaxlib>=0.1.47",
     # https://github.com/google/jax/discussions/9951#discussioncomment-3017784
```
Swapped the commend and jaxopt lines

bamos commented on src/ott/solvers/nn/conjugate_solver.py:
```diff
@@ -0,0 +1,57 @@
+
+from jaxopt import LBFGS
+
+ConjugateResults = namedtuple("ConjugateResults", "val grad num_iter")
```
Updated

michalk8 commented on docs/notebooks/neural_dual.ipynb:
```diff
@@ -12,7 +12,7 @@
    "metadata": {},
```
<pre class="ql-syntax" spellcheck="false">Line #5.    from functools import partial
</pre><p>Use <code>import functools</code> and <code>import dataclasses </code>rather than <code>from ...</code> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/219/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

michalk8 replied:
<pre class="ql-syntax" spellcheck="false">Line #1.    def plot_ot_map(neural_dual, source, target, inverse=False):
</pre><p>For consistency, would rename <code>inverse=False -&gt; forward=True</code>. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/219/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

michalk8 replied:
<p>The potentials are swapped, should be <code>\nabla f(source)</code>and <code>\nabla g(target)</code> to match the changes made in the code.</p><p> </p><p> </p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/219/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

michalk8 commented on docs/notebooks/neural_dual.ipynb:
```diff
@@ -12,7 +12,7 @@
    "metadata": {},
```
<pre class="ql-syntax" spellcheck="false">Line #1.    def get_optimizer(optimizer, lr, b1, b2, eps):
</pre><p>I don't think <code>get_optimizer</code> is  currently used, would remove and rename <code>get_optimizer_decay</code> to <code>get_optimizer</code> with <code>AdamW</code> as the only option (+ add defaults for <code>b1/b2</code>).    </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/219/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

bamos replied:
I'm still tweaking the optimizer around and will keep both for now. Once it's settled I'm also thinking about just inlining the optimizer

bamos replied:
Inlined

michalk8 commented on docs/notebooks/neural_dual.ipynb:
```diff
@@ -12,7 +12,7 @@
    "metadata": {},
```
<p>This text need to be adapted to mention the conjugate solver - I'd mention both that <code>g</code> can be <code>ICNN</code> or directly the potential <code>MLP</code> .</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/219/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

bamos replied:
<p>Will add in the next significant pass</p>

michalk8 commented on docs/notebooks/neural_dual.ipynb:
```diff
@@ -12,7 +12,7 @@
    "metadata": {},
```
<p>Maybe it would be worthwhile to also keep the previous approach where <code>neural_g</code> is an <code>ICNN</code> , e.g., as:</p><p><code>num_train_iters = 10_000</code></p><p><code>g_as_icnn = False</code></p><p> </p><p><code>neural_f = icnn.ICNN(...)</code></p><p><code>if g_as_icnn:</code></p><p><code>    neural_g = icnn.ICNN(...)</code></p><p><code>else:</code></p><p><code>    neural_g = mlp.PotentialMLP(...)</code></p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ott-jax/ott/pull/219/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

bamos replied:
<p>I think switching back to an ICNN for <code>g</code> doesn't work very well right now. I can debug a bit when trying to fix the inverse map and see if I can get it doing something reasonable </p>

bamos commented on src/ott/solvers/nn/conjugate_solver.py:
```diff
@@ -0,0 +1,57 @@
+ConjugateResults = namedtuple("ConjugateResults", "val grad num_iter")
+
+
+@dataclass
```
Added the decorator

bamos commented on src/ott/solvers/nn/conjugate_solver.py:
```diff
@@ -0,0 +1,57 @@
+  decrease_factor: float = 2. / 3.
+  ls_method: str = 'strong-wolfe'
+
+  def conj_min_obj(self, x, f, y):
```
Replaced with `conj_min_obj = lambda x: f(x) - x.dot(y)` inline in `solve`

bamos commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -108,55 +146,135 @@ def setup(
+        f"`W2NeuralDual` are not consistent. Proceeding with " \
+        f"the `W2NeuralDual` setting, with positive weights " \
+        f"being {self.pos_weights}."
+    if isinstance(neural_f, icnn.ICNN):
```
Merged, and updated to `is not`. Keeping unresolved for now to check

bamos commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -108,55 +146,135 @@ def setup(
+
+    for step in tqdm(range(self.num_train_iters)):
+      # execute training steps
+      train_batch["source"] = jnp.array(next(trainloader_source))
```
Updated here and everywhere else in the file

bamos commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -108,55 +146,135 @@ def setup(
     return (res, logs) if self.logging else res
 
-  def train_neuraldual(
+  def train_neuraldual_parallel(
```
I originally decided to keep these separate since the conditionals necessary to merge them would decrease how easy it is to read/edit them, but maybe it's not so bad. I'll keep this unresolved for now and add it in a larger editing pass soon

bamos commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -64,14 +96,17 @@ def __init__(
       seed: int = 0,
       pos_weights: bool = True,
       beta: float = 1.0,
+      conjugate_solver: Optional[Callable] = default_conjugate_solver,
```
I just added an abstract class for the conjugate solver and added that here for the type

bamos commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -108,55 +146,135 @@ def setup(
-  ) -> Union[potentials.DualPotentials, Tuple[potentials.DualPotentials,
-                                              Train_t]]:
-    logs = self.train_neuraldual(
+      callback: Optional[Callable] = None,
```
Added `Callback_t = Callable[[int, potentials.DualPotentials], None]` and updated everywhere

bamos commented:
Thanks for the review @michalk8! I just pushed in https://github.com/ott-jax/ott/pull/219/commits/0bf2467ce8cba7ea166c72699f14bfd725e277e1 to integrate most of your smaller comments and left some larger ones unresolved for now that I'll come back to in a larger pass

marcocuturi reviewed (approved):
thanks a lot Brandon for this amazing contribution!!

michalk8 commented:
@bamos in #233 the parser in notebooks was changed to use `myst-nb` in order to provide a better way of specifying links to the classes/functions/etc., just wanted to let you know.

marcocuturi commented:
Hi @bamos, now that ICML is behind us do you want to push this ? do you need any help?

bamos commented:
> Hi @bamos, now that ICML is behind us do you want to push this?

Yes! I'll rebase on top of the main branch and see if I can fix the inverse map over the next few days. (I think it's a matter of finding the right hyper-parameters...) Then it should be ready for a final review/merge.

bamos commented:
Finally got it working! :)

![image](https://user-images.githubusercontent.com/707462/217433859-e2a673fc-67f2-42f8-88e2-5c9b9c978bbf.png)
![image](https://user-images.githubusercontent.com/707462/217433866-3a84dd9b-12f7-4fb8-bdf3-25f1fa0e9c4e.png)

---

The key to making these work was to use a non-convex MLP for the potential instead of the ICNNs. Unfortunately the ICNNs were extremely difficult to debug in this setting and I couldn't figure out how to get the inverse map working correctly. (It seems like even numerically solving for the inverse still results in a lot of artifacts, so it may indicate that the forward potential with the ICNN is also off a little although it visually looks ok)

I also added [a `back_and_forth` option](https://github.com/ott-jax/ott/blob/e59563224d8e0a21c67f67c4bf54714a2656fc6f/src/ott/solvers/nn/neuraldual.py#L252), inspired from [this paper](https://www.math.purdue.edu/~jacob225/papers/fast_ot.pdf) that alternates between updating the forward and inverse maps. This helps stabilize the potentials to be conjugates of each other. (This also came up in a conversation I was having with @APooladian.) Without this, the non-convex potentials quickly diverge: 

![image](https://user-images.githubusercontent.com/707462/217435498-3f69b73f-a98d-4d9e-839c-c7d4c3738686.png)

Tomorrow I'll make a pass through the notebook (it's not updated with these results) and should be able to address the remaining TODOs here to have a final version of this PR ready for review. I'll make the notebook start with the simpler task that the ICNN can fit, and then afterwards show how an MLP can be swapped in and used for this harder setting.

bamos commented:
The GPU tests failed [here](https://github.com/ott-jax/ott/actions/runs/4131372214/jobs/7138941251) with error

```
ERROR tests/utils_test.py - AttributeError: module 'jax.sharding' has no attribute 'PartitionSpec'
```

Flax just released [v0.6.5](https://github.com/google/flax/releases/tag/v0.6.5) a few hours ago that I think is causing this. It could either be fixed by bumping up the jax version or upper-bounding the flax version. I just added https://github.com/ott-jax/ott/pull/219/commits/839719e8644eee05d8de7f3d8b3bc646ac3805a2 to try to fix this, but can revert/edit if you have a better solution.

bamos commented:
Hi @michalk8 @marcocuturi, I've a version ready to be reviewed and merged in! Can you please go through the latest version and let me know if there's anything else you want me to get in? I tried to re-request reviews from both of you but for some reason github only lets me send it for one of you.

bamos commented:
Also @marcocuturi, I tried the idea you've mentioned before of starting with the ICNN weights negative and annealing them to be positive by shifting them throughout training, i.e., `z_weights = softplus(x+offset)-offset` where `offset` started at 10 and decreased to 0 over a few thousand training iterations. I wasn't able to quickly make it work nicely, as it seems like the model started relying too much on the negative values and wasn't able to correct them as they were pushed to be positive, although there may be a better initialization/annealing strategy than what I tried. I don't have the code easily ready to keep building on this as I hacked in something to condition the positive dense layer on the training iteration, but it could be relatively easy to add back if a step-dependent annealing seems promising elsewhere.

michalk8 commented:
> The GPU tests failed [here](https://github.com/ott-jax/ott/actions/runs/4131372214/jobs/7138941251) with error
> 
> ```
> ERROR tests/utils_test.py - AttributeError: module 'jax.sharding' has no attribute 'PartitionSpec'
> ```
> 
> Flax just released [v0.6.5](https://github.com/google/flax/releases/tag/v0.6.5) a few hours ago that I think is causing this. It could either be fixed by bumping up the jax version or upper-bounding the flax version. I just added [839719e](https://github.com/ott-jax/ott/commit/839719e8644eee05d8de7f3d8b3bc646ac3805a2) to try to fix this, but can revert/edit if you prefer something else

Seems like new `flax` requires `jax>=0.4.2` [here](https://github.com/google/flax/blob/main/setup.py#L29), whereas our GPU CI needs `jax<0.4` because of this issue which I wasn't able to resolve so far: https://github.com/google/jax/issues/13758; will take another look at it.

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -96,66 +153,195 @@ def __init__(
       trainloader_target: Iterable[jnp.ndarray],
       validloader_source: Iterable[jnp.ndarray],
       validloader_target: Iterable[jnp.ndarray],
+      callback: Optional[Callback_t] = None,
```
Missing type hints of the callback.

bamos replied:
It's hinting to `Optional[Callback_t]`, is there something else I should add?

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -96,66 +153,195 @@ def __init__(
+    }
+    valid_logs = {"valid_loss_f": [], "valid_loss_g": [], "valid_w_dist": []}
+
+    for step in tqdm(range(self.num_train_iters)):
```
I'd assign a variable here `update_forward = step % 2 = 0` to be explicit.

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -96,66 +153,195 @@ def __init__(
+            train_batch,
+        )
+
+        if self.logging and step % self.log_freq == 0:
```
There's too much code duplication when updating the logs, would move it outside the `if` as
```python
train_logs["train_loss_f"].append(float(loss_f))
train_logs["train_loss_g"].append(float(loss_g))
train_logs["train_w_dist"].append(float(w_dist))
train_logs["directions"].append('forward' if update_forward else 'backward')
```

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -12,48 +12,95 @@
 import optax
 from flax import core
+from flax.core.frozen_dict import FrozenDict
+from jax.lax import stop_gradient
```
Would not import and use `jax.lax.stop_gradient` directly

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -219,65 +408,93 @@ def train_neuraldual(
-
-      s_dot_grad_g_s = jnp.sum(source * grad_g_s, axis=1)
+      if self.g_returns_potential:
+        g_grad = jax.vmap(
```
For clarify, I think this could be a method of `PotentialMixin` as
```python
def grad(self, params) -> Callable[[jnp.ndarray], ndarray]:
   if self.is_potential:
       return jax.vmap(lambda y: jax.grad(self, ...))
   return self  # __call__ already defined a gradient
```

bamos replied:
That is nice! And the potential's objective value can be put into a similar interface instead of inlined as I've done in `to_dual_potentials`.

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -299,14 +527,59 @@ def step_fn(state_f, state_g, batch):
     f = lambda x: self.state_f.apply_fn({"params": self.state_f.params}, x)
-    g = lambda x: self.state_g.apply_fn({"params": self.state_g.params}, x)
+
+    def g_prediction(y):
```
Missing type hints.

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -299,14 +527,59 @@ def step_fn(state_f, state_g, batch):
+        g_y = -f(grad_g_y) + jax.vmap(jnp.dot)(grad_g_y, y)
+        return g_y.squeeze(0) if squeeze else g_y
+
+    def g_finetuned(y):
```
Missing type hints.

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -327,3 +600,124 @@ def _penalize_weights_icnn(params: Dict[str, jnp.ndarray]) -> float:
     return penalty
+
+
+def plot_ot_map(
```
I think this should belong to `potentials.py`

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -327,3 +600,124 @@ def _penalize_weights_icnn(params: Dict[str, jnp.ndarray]) -> float:
+  fig = plt.figure(facecolor="white")
+  ax = fig.add_subplot(111)
+
+  if not inverse:
```
This should be simplified as
```python
if not inverse:
   color_src, color_tgt = ..., ...
   label_src, label_tgt = ..., ...
   label_transport = ....\
else:
   ...
ax.scatter(...)
```

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -327,3 +600,124 @@ def _penalize_weights_icnn(params: Dict[str, jnp.ndarray]) -> float:
+    learned_potentials: potentials.DualPotentials,
+    source: jnp.ndarray,
+    target: jnp.ndarray,
+    inverse: bool = False
```
Expose `**kwargs: Any` to `ax.scatter` please.

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -327,3 +600,124 @@ def _penalize_weights_icnn(params: Dict[str, jnp.ndarray]) -> float:
+        label=r"$\nabla g(target)$",
+    )
+
+  fig.legend(ncol=3, loc="upper center")
```
`ax.legend`

bamos replied:
`fig.legend` puts it above the axis:

![image](https://user-images.githubusercontent.com/707462/217826706-0eb8b2dc-71f0-4e82-8334-d6f531c13104.png)

but `ax.legend` puts it inside:

![image](https://user-images.githubusercontent.com/707462/217828658-74c64f5d-7238-45de-a69a-ac1a59b987df.png)

I'll keep `fig.legend` unless you meant that you prefer to have it inside

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -327,3 +600,124 @@ def _penalize_weights_icnn(params: Dict[str, jnp.ndarray]) -> float:
+    source: jnp.ndarray,
+    target: jnp.ndarray,
+    inverse: bool = False
+) -> Tuple[matplotlib.figure.Figure, matplotlib.axes.Axes]:
```
Please add `matplotlib` to the [intersphinx_mapping](https://github.com/ott-jax/ott/blob/main/docs/conf.py#L68).

bamos replied:
added `"matplotlib": ('https://matplotlib.org/stable/', None)`

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -327,3 +600,124 @@ def _penalize_weights_icnn(params: Dict[str, jnp.ndarray]) -> float:
+  """
+
+  def draw_arrows(a, b):
+    plt.arrow(
```
Should be `ax.arrow`

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -327,3 +600,124 @@ def _penalize_weights_icnn(params: Dict[str, jnp.ndarray]) -> float:
+  Returns: matplotlib figure and axis with the plots
+  """
+
+  def draw_arrows(a, b):
```
draw_arrows(a: np.ndarray, b: np.ndarray) -> None:

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -327,3 +600,124 @@ def _penalize_weights_icnn(params: Dict[str, jnp.ndarray]) -> float:
+def plot_potential(
+    learned_potentials: potentials.DualPotentials,
+    inverse: bool = False,
+    alpha: float = 0.05
```
Allow passing `**kwargs: Any` to `contourf`.

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -327,3 +600,124 @@ def _penalize_weights_icnn(params: Dict[str, jnp.ndarray]) -> float:
+  X2flat = np.ravel(X2)
+  X12flat = np.stack((X1flat, X2flat)).T
+  if not inverse:
+    Zflat = learned_potentials._f(X12flat)
```
Use the public `.f`, not `._f`.

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -299,14 +527,59 @@ def step_fn(state_f, state_g, batch):
+    r"""Return the Kantorovich dual potentials from the trained potentials.
+
+    If `g` returns the gradient of the g potential, \nabla_y g,
+    i.e. `g.returns_potential == False`,
```
Would remove this line, doesn't render nicely.

bamos replied:
Updated to:

![image](https://user-images.githubusercontent.com/707462/217833786-b07f5038-3349-4ea6-90d3-ecf3f8342347.png)

bamos commented on src/ott/solvers/nn/conjugate_solver.py:
```diff
@@ -0,0 +1,112 @@
+
+@dataclass
+@utils.register_pytree_node
+class FenchelConjugateLBFGS(FenchelConjugateSolver):
```
What should I update for this?

michalk8 replied:
Ah sorry, this is already a `dataclass` (because of `@utils.register_pytree_node`), the `@dataclass` can be removed then.

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -96,66 +153,195 @@ def __init__(
-    self.train_step_f = self.get_step_fn(train=True, to_optimize="f")
-    self.valid_step_f = self.get_step_fn(train=False, to_optimize="f")
+    # Assume g defines the potential unless this attribute is set.
+    self.g_returns_potential = getattr(neural_g, 'returns_potential', True)
```
This is related to this comment: https://github.com/ott-jax/ott/pull/219#discussion_r1101289186; would consider doing the ``isinstance(neural_g, PotentialMixin)`` check.

bamos replied:
It's much cleaner that way, thanks!

bamos commented on src/ott/solvers/nn/mlp.py:
```diff
@@ -0,0 +1,79 @@
+      output dimension of the last layer is automatically set to
+      1 (if `returns_potential=True`) or the dimension
+      of the input (if `returns_potential=False`).
+    returns_potential: Model the potential if `True`, otherwise
```
> I think is potential is more apt.

Do you mean renaming `returns_potential` to just `potential`, or `is_potential`?

> [The Mixin]

Good idea, should I add it as `neuraldual.PotentialMixin` or create a new module for it?

bamos replied:
I decided to make this a base class `PotentialBase` as it seems reasonable to require anybody defining their own potentials to subclass from this. I put it in `ott.solvers.nn.potential_base` for now and can move it

bamos commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -12,48 +12,95 @@
+Callback_t = Callable[[int, potentials.DualPotentials], None]
+Conj_t = Optional[FenchelConjugateSolver]
+
+DEFAULT_CONJUGATE_SOLVER = FenchelConjugateLBFGS(
```
Moving to the bottom there

bamos commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -327,3 +600,124 @@ def _penalize_weights_icnn(params: Dict[str, jnp.ndarray]) -> float:
+  else:
+    Zflat = np.array(jax.vmap(learned_potentials._g)(X12flat))
+
+    vmin, vmax = np.quantile(Zflat, alpha), np.quantile(Zflat, 1 - alpha)
```
Great spot!

bamos commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -327,3 +600,124 @@ def _penalize_weights_icnn(params: Dict[str, jnp.ndarray]) -> float:
+  if not inverse:
+    Zflat = learned_potentials._f(X12flat)
+  else:
+    Zflat = np.array(jax.vmap(learned_potentials._g)(X12flat))
```
Updated to be more consistent

bamos commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -299,14 +527,59 @@ def step_fn(state_f, state_g, batch):
+    .. math::
+      g(y) = -f(\nabla_y g(y)) + y^T \nabla_y g(y)
+
+    where :math:`\nabla_y g(y)` is detached for the envelope theorem
```
Added here and where the Sinkhorn solver references the envelope theorem: https://github.com/ott-jax/ott/blob/116dbbf025ee6cce395fbb68490cd99bb549d664/src/ott/solvers/linear/sinkhorn.py#L946

michalk8 commented on pyproject.toml:
```diff
@@ -15,10 +15,11 @@ authors = [
     "numpy>=1.18.4, !=1.23.0",
     "matplotlib>=3.0.0",
-    "flax>=0.5.2",
+    "flax>=0.5.2,<=0.6.4",
```
I'd remove this pin and pin jax/flax in the GPU CI, see [here](https://github.com/ott-jax/ott/pull/244/commits/64ab7e7ddcaa5dc2e6476237532806237ea6c454).
Problem is that the server has old GPUs that need old drivers and I can't upgrade them at the moment.

bamos replied:
Ok, reverted

michalk8 commented on src/ott/problems/linear/potentials.py:
```diff
@@ -258,3 +264,125 @@ def epsilon(self) -> float:
     return [self._f, self._g, self._prob, self._f_xx, self._g_yy], {}
+
+
+def plot_ot_map(
```
Think this should be a method of the `DualPotentials`, since it takes it as a first argument (same for the other function).

bamos replied:
That makes sense, moved there

michalk8 commented on src/ott/problems/linear/potentials.py:
```diff
@@ -258,3 +264,125 @@ def epsilon(self) -> float:
+    target: samples from the target measure
+    inverse: if the inverse map from the potentials
+      should be used
+    ax: axis to add the plot to
```
Missing docstring for `**kwargs`.

michalk8 commented on src/ott/problems/linear/potentials.py:
```diff
@@ -258,3 +264,125 @@ def epsilon(self) -> float:
+    inverse: if the inverse map from the potentials
+      should be used
+    alpha: Quantile to filter the potentials with
+    ax: axis to add the plot to
```
Missing docstring for `**kwargs`.

michalk8 commented on src/ott/problems/linear/potentials.py:
```diff
@@ -32,7 +38,7 @@ class DualPotentials:
   Args:
     f: The first dual potential function.
-    g: The second dual potential function.
+    The: g second dual potential function.
```
Undo this please.

michalk8 commented on src/ott/solvers/linear/sinkhorn.py:
```diff
@@ -944,6 +944,7 @@ def output_from_state(
       When differentiating the regularized OT cost, and assuming Sinkhorn has
       run to convergence, Danskin's (or the envelope)
       `theorem <https://en.wikipedia.org/wiki/Danskin%27s_theorem>`_
+      :cite:`danskin:67,bertsekas:71`
```
Thanks, this is great!

michalk8 commented on src/ott/solvers/nn/conjugate_solver.py:
```diff
@@ -0,0 +1,118 @@
+from ott import utils
+
+__all__ = [
+    "ConjugateResults", "FenchelConjugateSolver", "FenchelConjugateLBFGS"
```
The `DEFAULT_CONJUGATE_SOLVER` should also be here.

michalk8 commented on src/ott/problems/linear/potentials.py:
```diff
@@ -258,3 +264,125 @@ def epsilon(self) -> float:
+  else:
+    fig = ax.get_figure()
+
+  x1 = np.linspace(-6, 6)
```
Think it would be great if user had more control over the `x1/x2` by allowing them to pass this explicitly.

bamos replied:
Added `bounds` as an argument to the function, along with `num_grid` to control how many points each linspace here uses

michalk8 commented on src/ott/problems/linear/potentials.py:
```diff
@@ -258,3 +264,125 @@ def epsilon(self) -> float:
+  X1, X2 = np.meshgrid(x1, x2)
+  X1flat = np.ravel(X1)
+  X2flat = np.ravel(X2)
+  X12flat = np.stack((X1flat, X2flat)).T
```
Would call `jnp.{h,v}stack` to explicitly convert to `jax` + no need to do the transpose.

bamos replied:
Replaced with `X12flat = jnp.hstack((X1.reshape(-1, 1), X2.reshape(-1, 1)))`

bamos replied:
And sampled from the linspace with jax arrays

michalk8 commented on src/ott/problems/linear/potentials.py:
```diff
@@ -258,3 +264,125 @@ def epsilon(self) -> float:
+
+  CS = ax.contourf(X1, X2, Z, cmap="Blues", **kwargs)
+  fig.colorbar(CS, ax=ax)
+  fig.tight_layout()
```
I'd do this only when `ax=None`.

michalk8 commented on src/ott/solvers/nn/potential_base.py:
```diff
@@ -0,0 +1,113 @@
+    if self.is_potential:
+      return self.apply({'params': params}, x)
+    else:
+      if self.is_potential:
```
This is always false.

bamos replied:
Fixed (it was an accidental copy/paste)

michalk8 commented on src/ott/solvers/nn/potential_base.py:
```diff
@@ -0,0 +1,113 @@
+  def potential_gradient(
+      self, params: Optional[frozen_dict.FrozenDict[str, jnp.ndarray]],
+      x: jnp.ndarray
+  ):
```
-> jnp.ndarray:

bamos commented:
(Just pushed the re-run notebook in)

michalk8 commented on src/ott/solvers/nn/potential_base.py:
```diff
@@ -0,0 +1,113 @@
+      if it defined the gradient.
+    """
+
+  def potential_value(
```
I wonder whether this should also evaluate it on `x` or just return the potential function, as I think it will be more jax-y; would return `Callable[[jnp.ndarray], jnp.ndarray]`.

bamos replied:
I made it behave similar to the `apply` function, which also evaluates the point. This way they're stored along in the training state in the same way as the `apply` function rather than needing an additional call to create the function. Do you think that's reasonable?

bamos replied:
Actually I thought about it some more and it seems a little nicer to make them functions. I just pushed this in https://github.com/ott-jax/ott/pull/219/commits/ce90b67ff4f755ad664bf4deede78584717160fb and will keep this thread open in case you have any other thoughts here

michalk8 commented on src/ott/solvers/nn/potential_base.py:
```diff
@@ -0,0 +1,113 @@
+      The gradient of the potential
+    """
+    if self.is_potential:
+      return jax.vmap(
```
In principle if `potential_value` were to become a `potential_fn`, you could then do
`return jax.vmap(jax.grad(self.potential_fn, argnums=...))` and rename this to `grad_potential_fn` + remove `x`.

michalk8 replied:
Also helps with the ugly `yapf` formatting 😅 .

bamos replied:
That is nice, updated to this:
![image](https://user-images.githubusercontent.com/707462/217894483-9ec1ca1f-4247-4812-8518-664e40540094.png)

michalk8 commented on src/ott/problems/linear/potentials.py:
```diff
@@ -258,3 +264,125 @@ def epsilon(self) -> float:
+      source[:, 0],
+      source[:, 1],
+      color="#1A254B",
+      alpha=0.5,
```
Can we please expose `alpha: float = 0.5`?

michalk8 commented on src/ott/problems/linear/potentials.py:
```diff
@@ -258,3 +264,125 @@ def epsilon(self) -> float:
+      **kwargs,
+  )
+
+  fig.legend(ncol=3, loc="upper center")
```
Re your comment, I think I'd prefer the legend inside with defaults (no `ncol`, etc.). If you want, you can add `legend_kwargs: Mapping[str, Any]` for more control.

bamos replied:
I don't like how it moves around the placement of the legend depending on the contents, are you okay with it?

![image](https://user-images.githubusercontent.com/707462/217888336-ba8d3838-8a64-4ce3-83bf-90ad83fc8e9f.png)

michalk8 replied:
I don't mind this, but would re-consider exposing the legend kwargs.

bamos replied:
Hmm, are you ok with `**kwargs` of `plot_ot_map` being the kwargs for scatter and having `legend_kwargs` separate? We could also make them both dicts as `legend_kwargs` and `scatter_kwargs`.

michalk8 replied:
Both sound good, will leave it up to you.

michalk8 commented on src/ott/problems/linear/potentials.py:
```diff
@@ -258,3 +264,125 @@ def epsilon(self) -> float:
+  fig.legend(ncol=3, loc="upper center")
+
+  for i in range(source.shape[0]):
+    draw_arrows(source[i, :], grad_state_s[i, :])
```
Think at this point, there's no need to have the local function above (it's just one line), but will leave it up to you.

bamos replied:
Inlined

michalk8 commented on src/ott/problems/linear/potentials.py:
```diff
@@ -258,3 +264,125 @@ def epsilon(self) -> float:
+  for i in range(source.shape[0]):
+    draw_arrows(source[i, :], grad_state_s[i, :])
+
+  return fig, ax
```
Consider returning just the `ax`, since user can always do `ax.get_figure()`.

bamos replied:
It is a nice convenience to just do this in the notebooks/everywhere else it's called instead of manually calling `ax.get_figure()` everywhere:

![image](https://user-images.githubusercontent.com/707462/217888852-fd1ac61a-4a9e-4b1a-9d3a-62a28f3ff44e.png)

michalk8 replied:
Ok, let's return both fig and ax.

michalk8 commented on src/ott/solvers/nn/conjugate_solver.py:
```diff
@@ -0,0 +1,118 @@
+
+  def solve(
+      self,
+      f: Callable,
```
`f: Callable[[jnp.ndarray], float]`.

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -12,73 +12,114 @@
-
-  Original algorithm is described in :cite:`makkuva:20`.
+Callback_t = Callable[[int, potentials.DualPotentials], None]
+Conj_t = Optional[conjugate_solver.FenchelConjugateSolver]
```
No need for this alias, used only once.

bamos replied:
I originally did this because unfortunately inlining it seems to create a bug between `autoflake` and `flake8`: `autoflake` formats it to:

![image](https://user-images.githubusercontent.com/707462/217891496-4946fd2e-b0a7-4212-a8fa-5eaaca38601b.png)

but then flake gives the error:

![image](https://user-images.githubusercontent.com/707462/217891535-50d40e07-94da-406e-bf3d-5c63cf180b1f.png)

And any manual edits I do to the line get automatically reverted with `autoflake`. Do you have any other ideas to fix it?

michalk8 replied:
Ok, let's keep it as it is then!

michalk8 commented on docs/conf.py:
```diff
@@ -66,6 +66,8 @@
     "scikit-sparse": ("https://scikit-sparse.readthedocs.io/en/latest/", None),
     "scipy": ("https://docs.scipy.org/doc/scipy/reference/", None),
     "pot": ("https://pythonot.github.io/", None),
+    "jaxopt": ("https://jaxopt.github.io/", None),
```
The `BFGS` reference from `jaxopt` doesn't render, can you please try https://jaxopt.github.io/stable/?

bamos replied:
That fixed it:

![image](https://user-images.githubusercontent.com/707462/217890463-75858094-18e0-45a6-b294-6664c9b6604e.png)

michalk8 commented on docs/solvers/nn.rst:
```diff
@@ -10,12 +10,25 @@ Neural Dual
     :toctree: _autosummary
 
-    icnn.ICNN
+    potential_base.PotentialBase
```
I think this module is a bit too fragmented, would try to consolidate the classes into fewer files, e.g.:
```rst
.. autosummary::
    :toctree: _autosummary

    potentials.PotentialBase  # not sure if necessay
    potentials.ICNN
    potentials.MLP
    layers.PositiveDense
    layers.PosDefPotentials  # we might need to rename it + improve docstring
```
Am open to suggestions, also pinging @marcocuturi

bamos replied:
I'm open to putting the base, ICNN, MLP in the same class. Also regardless of that, is it somewhat confusing to have modules/objects approximately called potentials in `ott.problems.linear` and `ott.solvers.nn`?

marcocuturi replied:
I agree with Michal on consolidation, and I think bundling MLP / ICNN and base makes sense indeed Brandon!

I do find it a bit confusing to have 2 references to potentials.

What I also find a bit confusing is calling `potential` something that's just a gradient of something (that may not be a potential).

Maybe call this simply `NN` or `Base` or `NNBase` ? Then from context people figure out whether it's a potential or something that looks like a map.

bamos replied:
What do you think if I create `ott.solvers.nn.models` and then refactor `nn.potential_base.PotentialBase` into `nn.models.NeuralModelBase`, and also put `ICNN` and `MLP` in there as subclasses?

bamos replied:
Just consolidated in https://github.com/ott-jax/ott/pull/219/commits/d7a948c56a923d4144eb66948c77257e24c8a543 and will push the updated notebooks soon once they finish running. Leaving this open in case there are any remaining changes you all want for this

bamos commented:
I just went through the latest batch of comments and pushed in some commits addressing them. I've also rebased with `upstream/main`, fixed the merge conflicts, and force-pushed. I pushed a version of the notebook without outputs and am doing a final run of that now.

marcocuturi commented on src/ott/problems/linear/potentials.py:
```diff
@@ -157,6 +161,139 @@ def tree_unflatten(
   ) -> "DualPotentials":
     return cls(*children, **aux_data)
 
+  def plot_ot_map(
```
can we imagine putting this in the tools section of the code? feels a bit out of place in this context, or is there a good reason to have it inside the linear problems area, WDYT?

bamos replied:
Yeah, I'm okay with them over there with a name like `plot_W2_{map,potentials}`. How about you @michalk8?

michalk8 replied:
Yes, `ott/tools/potentials.py` sounds good. As for renaming, don't think we should add the W2; the `plot_ot_map/plot_potential` is ok.
Also for consistency, I'd rename `inverse` argument to `forward`.

bamos replied:
What about `plot_neural_ot_map` and `plot_neural_potential` to clarify it's not for the discrete/sinkhorn maps?

michalk8 replied:
These methods are available to the `EntropicMap`, therefore `neural` won't work.

bamos replied:
Ah, then `plot_contunous_ot_map` and `plot_continuous_potential`? Or do you want to keep the shorter names?

marcocuturi replied:
maybe continuous is not needed and shorter is ok?

bamos replied:
Ok, adding that now

marcocuturi commented on src/ott/solvers/nn/potential_base.py:
```diff
@@ -0,0 +1,106 @@
+
+    Returns:
+      ``True`` if the module defines the potential's value, ``False``
+      if it defined the gradient.
```
defines

marcocuturi reviewed (commented):
Thanks a lot Brandon!! I think we're almost good up to minor renaming!

michalk8 commented on src/ott/problems/linear/potentials.py:
```diff
@@ -157,6 +161,141 @@ def tree_unflatten(
+        should be used
+      ax: axis to add the plot to
+      alpha: transparency of the scatter plot points
+      kwargs: additional kwargs passed into ``scatter``
```
Use :meth:\`~matplotlib.axes.Axes.scatter\`.

michalk8 commented on src/ott/problems/linear/potentials.py:
```diff
@@ -157,6 +161,141 @@ def tree_unflatten(
+    """Plot data and learned optimal transport map.
+
+    Args:
+      learned_potentials: the potentials object for the map
```
This argument is unused.

michalk8 commented on src/ott/problems/linear/potentials.py:
```diff
@@ -157,6 +161,141 @@ def tree_unflatten(
+      inverse: bool = False,
+      alpha: float = 0.05,
+      ax: Optional[matplotlib.axes.Axes] = None,
+      bounds: Tuple[float] = (-6, 6, -6, 6),
```
Think it's a bit nicer either to have 2 arguments ({x,y}_bounds), or have `)(-6, 6), (-6, 6))` instead of a flat tuple.

bamos replied:
Updated to `{x,y}_bounds`

michalk8 commented on src/ott/problems/linear/potentials.py:
```diff
@@ -157,6 +161,141 @@ def tree_unflatten(
+      alpha: float = 0.05,
+      ax: Optional[matplotlib.axes.Axes] = None,
+      bounds: Tuple[float] = (-6, 6, -6, 6),
+      num_grid: float = 50,
```
Should be an int.

michalk8 commented on docs/solvers/nn.rst:
```diff
@@ -10,12 +10,25 @@ Neural Dual
 
-    neuraldual.NeuralDualSolver
+    neuraldual.W2NeuralDual
+    neuraldual.plot_ot_map
```
These are now part of the `DualPotentials`, please remove.

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -94,68 +135,191 @@ def __init__(
+        )
+
+        if self.logging:
+          # log training progress
```
Call `self._update_logs`.

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -212,101 +377,158 @@ def train_neuraldual(
+
+    Args:
+      finetune_g: Run the conjugate solver to finetune the prediction.
+    """
```
Missing `Returns:`

bamos commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -212,101 +377,158 @@ def train_neuraldual(
+    if finetune_g:
+      g_prediction = g
+
+      def g_finetuned(y: jnp.ndarray) -> jnp.ndarray:
```
Updating to:

```python
    f_value = self.state_f.potential_value_fn(self.state_f.params)                  
    g_value_prediction = self.state_g.potential_value_fn(self.state_g.params, f)    
                                                                                    
    def g_value_finetuned(y: jnp.ndarray) -> jnp.ndarray:                           
      x_hat = jax.grad(g_value_prediction)(y)                                       
      grad_g_y = jax.lax.stop_gradient(                                             
          self.conjugate_solver.solve(f, y, x_init=x_hat).grad                      
      )                                                                             
      g_y = -f(grad_g_y) + jnp.dot(grad_g_y, y)                                     
      return g_y                                                                    
                                                                                    
    return potentials.DualPotentials(                                               
      f=f_value,                                                                    
      g=g_value if not finetune_g else g_value_finetuned,                           
      cost_fn=costs.SqEuclidean(),                                                  
      corr=True                                                                     
    )
```

bamos commented:
I noticed that `ott.tools.plot` isn't available in the docs because it's not listed [in tools.rst](https://github.com/ott-jax/ott/blob/main/docs/tools.rst). Do you want to add the docs for at least the new plotting code here or keep it disabled?

bamos commented:
After this PR, I also think it would now be easy to evaluate OTT's solver here on [this Wasserstein-2 benchmark](https://arxiv.org/abs/2106.01954). It should give relatively competitive results and could provide a nice starting point for other methodological development. I usually run this benchmark on a cluster with hydra so maybe after this PR is merged I can see if I can easily create a new repo for running this benchmark with OTT. Beyond this it also seems interesting to add other neural W2 benchmark tasks (like some of the biology/cell/treatment ones), or to add other kinds of neural OT. It seems nice for the research community to have a collection of benchmark tasks with an easy and reproducible way of solving them with methods in OTT

bamos commented:
> After this PR, I also think it would now be easy to evaluate OTT's solver here on [this Wasserstein-2 benchmark](https://arxiv.org/abs/2106.01954). It should give relatively competitive results and could provide a nice starting point for other methodological development. I usually run this benchmark on a cluster with hydra so maybe after this PR is merged I can see if I can easily create a new repo for running this benchmark with OTT. Beyond this it also seems interesting to add other neural W2 benchmark tasks (like some of the biology/cell/treatment ones), or to add other kinds of neural OT. It seems nice for the research community to have a collection of benchmark tasks with an easy and reproducible way of solving them with methods in OTT

I've also noticed a few papers coming up that report the W2 on some of the 2D datasets and compare to their re-implementation of Makkuva's ICNN-based neural/minimax OT as a baseline. I think these baselines are underperforming compared to what is possible due to the implementation challenges/instabilities of neural OT that need to be carefully tuned/debugged (e.g., they're not using the conjugate finetuning). It could be useful to help standardize W2 computations on these with a well-tuned neural OT/static dual potentials baseline from OTT.

From [Neural Conservation Laws](https://arxiv.org/pdf/2210.01741.pdf):
![image](https://user-images.githubusercontent.com/707462/217997685-f844bab2-4209-4437-8deb-91b41153bad5.png)

From [OT Conditional Flow Matching](https://arxiv.org/pdf/2302.00482.pdf):
![image](https://user-images.githubusercontent.com/707462/217997718-d0d1dc7e-0c3f-4482-8bbe-4fad53642c61.png)

michalk8 commented on src/ott/solvers/nn/models.py:
```diff
@@ -0,0 +1,351 @@
+  ) -> PotentialValueFn_t:
+    r"""Return a function giving the value of the potential.
+
+    Applies the module if ``self.is_potential`` is ``True``, otherwise
```
:attr:\`is_potential\`

michalk8 commented on src/ott/solvers/nn/models.py:
```diff
@@ -0,0 +1,351 @@
+      params: parameters of the module
+      x: point to evaluate the value at
+      other_potential_value: function giving the value of the other potential.
+        Only needed when ``self.is_potential`` is ``False``.
```
:attr:\`is_potential\`

michalk8 commented on src/ott/solvers/nn/models.py:
```diff
@@ -0,0 +1,351 @@
+    if self.is_potential:
+      return lambda x: self.apply({"params": params}, x)
+    else:
+      assert other_potential_value_fn is not None
```
Can you please add some short message?

michalk8 commented on src/ott/solvers/nn/models.py:
```diff
@@ -0,0 +1,351 @@
+    else:
+      assert other_potential_value_fn is not None
+
+      def value_fn(x):
```
Missing type hints (sorry for being pesky, want to enable mypy in the future).

michalk8 commented on src/ott/solvers/nn/models.py:
```diff
@@ -0,0 +1,351 @@
+from jax.nn import initializers
+
+from ott.math import matrix_square_root
+from ott.solvers.nn.layers import PosDefPotentials, PositiveDense
```
For consistency, use `from ott.solvers.nn import layers`.

michalk8 commented on src/ott/tools/plot.py:
```diff
@@ -224,3 +227,149 @@ def barycentric_projections(
+    forward: bool = True,
+    alpha: float = 0.05,
+    ax: Optional[matplotlib.axes.Axes] = None,
+    x_bounds: Tuple[float] = (-6, 6),
```
`Tuple[float, float]`

michalk8 commented on src/ott/tools/plot.py:
```diff
@@ -224,3 +227,149 @@ def barycentric_projections(
   return _barycenters(ax, geom.y, arg.a, arg.b, arg.matrix, **kwargs)
+
+
+def plot_ot_map(
```
Based on this [conversation](https://github.com/ott-jax/ott/pull/219#discussion_r1101940346), I thought these will become methods of `DualPotentials` and will move `potentials.py` from `ott/problems/linear/potentials.py` to `ott/tools/potentials.py`.

michalk8 replied:
I'm also with leaving the file in `ott/problems/linear/potentials.py`, the reason why it is `linear` is because we currently don't have notion of potentials for GW, pinging @marcocuturi again.

bamos replied:
I accidentally overlooked the part where you suggested to move all of `potentials.py` to tools, I thought @marcocuturi  just meant to move the plotting code into `ott/tools/plot`. I'm ok either way

michalk8 replied:
I strongly prefer this as a method of `DualPotentials`, makes more sense.

As for
> I noticed that ott.tools.plot isn't available in the docs because it's not listed [in tools.rst](https://github.com/ott-jax/ott/blob/main/docs/tools.rst). Do you want to add the docs for at least the new plotting code here or keep it disabled?

I'd keep it disabled, as I plan to refactor this in the future.

michalk8 replied:
@bamos could you please move the plotting functions as methods of `DualPotentials`?

bamos replied:
Added back there: https://github.com/ott-jax/ott/pull/219/commits/e53c33907492d259b9fd7b6f9a03ee6ed70bd3e6

michalk8 commented:
> Hi @michalk8 @marcocuturi, I've a version ready to be reviewed and merged in! Can you please go through the latest version and let me know if there's anything else you want me to get in? I tried to re-request reviews from both of you but for some reason github only lets me send it for one of you.

Thanks @bamos , looks great. I left a few more comments, once they are resolved, we can merge.

michalk8 commented:
@bamos could you please also add some test for the `W2NeuralDual` using the potential MLP? Maybe parameterizing the [already present one](https://github.com/ott-jax/ott/blob/main/tests/solvers/nn/neuraldual_test.py#L80) will be the easiest.

bamos commented:
> @bamos could you please also add some test for the `W2NeuralDual` using the potential MLP? Maybe parameterizing the [already present one](https://github.com/ott-jax/ott/blob/main/tests/solvers/nn/neuraldual_test.py#L80) will be the easiest.

Yeah, I'll try adding every combination of supported networks to the tests; using ICNN, MLP (potential), and MLP (gradient). And can add some for the extra modes too (like `back_and_forth`)

bamos commented:
I noticed the test also has a `ToyDataset` class in there: https://github.com/ott-jax/ott/blob/main/tests/solvers/nn/neuraldual_test.py#L25 (also the `square_four` and `square_five` modes there are not used in the testing code). The source code for `ToyDataset` is duplicated in the notebooks, and in the main neural OT tutorial, I've updated it to sample batches. Do you think it makes sense to have this as a module in OTT to help reduce this duplication, or do you think datasets and sampling should remain separate from OTT's core? Connecting to https://github.com/ott-jax/ott/pull/219#issuecomment-1425149166 above it seems useful to have readily available data samplers for some common tasks somewhere (even outside of the core) that can easily be loaded and plugged into OTT's solvers

michalk8 commented:
> I noticed the test also has a `ToyDataset` class in there: https://github.com/ott-jax/ott/blob/main/tests/solvers/nn/neuraldual_test.py#L25 (also the `square_four` and `square_five` modes there are not used in the testing code). The source code for `ToyDataset` is duplicated in the notebooks, and in the main neural OT tutorial, I've updated it to sample batches. Do you think it makes sense to have this as a module in OTT to help reduce this duplication, or do you think datasets and sampling should remain separate from OTT's core? Connecting to [#219 (comment)](https://github.com/ott-jax/ott/pull/219#issuecomment-1425149166) above it seems useful to have readily available data samplers for some common tasks somewhere (even outside of the core) that can easily be loaded and plugged into OTT's solvers

I think having the toy example dataset in `ott/problems/nn/dataset.py` would be great to reduce code duplication, make it easier to try out neural OT + in the future, develop more complex interface (like converting from `LinearProblem`).

bamos commented:
> I think having the toy example dataset in `ott/problems/nn/dataset.py` would be great to reduce code duplication, make it easier to try out neural OT + in the future, develop more complex interface (like converting from `LinearProblem`).
---
> > @bamos could you please also add some test for the `W2NeuralDual` using the potential MLP? Maybe parameterizing the [already present one](https://github.com/ott-jax/ott/blob/main/tests/solvers/nn/neuraldual_test.py#L80) will be the easiest.
> 
> Yeah, I'll try adding every combination of supported networks to the tests; using ICNN, MLP (potential), and MLP (gradient). And can add some for the extra modes too (like `back_and_forth`)
----
I addressed both of these in https://github.com/ott-jax/ott/pull/219/commits/efa2507506c87a8900920b4f8a5dbee8acba6b2b

michalk8 commented on src/ott/problems/nn/dataset.py:
```diff
@@ -0,0 +1,155 @@
+  validloader_target: Iterable[jnp.ndarray]
+
+
+GAUSSIAN_CENTERS = {
```
I don't think this really needs a global constant, would just move to `__post_init__`.

michalk8 commented on src/ott/problems/nn/dataset.py:
```diff
@@ -0,0 +1,155 @@
+    name_target: name of the target measure
+    train_batch_size: the training batch size
+    valid_batch_size: the validation batch size
+    key: initial PRNG key
```
Missing Returns section.

michalk8 commented on src/ott/problems/nn/dataset.py:
```diff
@@ -0,0 +1,155 @@
+    name_source: Name_t,
+    name_target: Name_t,
+    train_batch_size: int = 1024,
+    valid_batch_size: int = 1000,
```
Any reason for `valid_batch_size` being different than train?

bamos replied:
Somebody could want to use a really large validation batch

bamos replied:
Ah but here they can both be 1024. I think before I had it set to 10000 and then later cut back to 1000

michalk8 replied:
1024 for both sounds great.

michalk8 commented on src/ott/problems/nn/dataset.py:
```diff
@@ -0,0 +1,155 @@
+                 'gaussian_square_four']
+
+
+class Dataloaders(NamedTuple):
```
I think slightly cleaner would be to call this `Dataset` and have only 2 properties:
`{source,target}_iter` + `create` return 2 datasets (1 for train, other for validation).

bamos replied:
I wrote it like this so that `create` returns an object that can be directly passed into the solver:

```
dataloaders = dataset.create('gaussian_simple', 'gaussian_circle')
learned_potentials = neural_dual_solver(*dataloaders)
```

But I agree it does seem slightly cleaner to group the source/target samplers together. Do you have a preference for how I modify it? We could also update the interface of the `neural_dual` to take a similar grouping

bamos replied:
```
train_dataloaders, val_dataloaders = dataset.create('gaussian_simple', 'gaussian_circle')
learned_potentials = neural_dual_solver(*train_dataloaders, *val_dataloaders)
```

michalk8 commented on src/ott/problems/nn/dataset.py:
```diff
@@ -0,0 +1,155 @@
+      )
+    self.centers = GAUSSIAN_CENTERS[self.name]
+
+  def __iter__(self):
```
Missing type hint.

michalk8 commented on src/ott/problems/nn/dataset.py:
```diff
@@ -0,0 +1,155 @@
+
+  def __post_init__(self):
+    if self.name not in GAUSSIAN_CENTERS:
+      raise ValueError(
```
Should be a `KeyError`.

michalk8 commented on src/ott/problems/nn/dataset.py:
```diff
@@ -0,0 +1,155 @@
+    raise ValueError('Only Gaussian datasets are supported')
+
+
+def create(
```
I'd rename this (or explain in docstring) this is toy/example data.

bamos replied:
I renamed it to `create_gaussian_mixture_samplers` (and if we add more data we can make something a little more general)

michalk8 commented on src/ott/problems/nn/dataset.py:
```diff
@@ -0,0 +1,155 @@
+      yield samples
+
+
+def get_sampler(name: str, **kwargs: Any) -> Iterable[jnp.ndarray]:
```
I don't think we need this function, `Name_t` can be `Literal['gaussian_square', 'gaussian_circle', ...]`.

bamos replied:
I removed the `gaussian_` prefix from here for now since they're all Gaussian and inlined `get_sampler` back in to get rid of it. Are you saying to remove `Name_t` as well?

michalk8 replied:
No, the `Name_t` alias is good, thank you!

michalk8 commented on src/ott/solvers/nn/models.py:
```diff
@@ -0,0 +1,352 @@
+  Args:
+    dim_hidden: sequence specifying size of hidden dimensions. The
+      output dimension of the last layer is automatically set to
+      1 (if `is_potential=True`) or the dimension
```
Would use double \`\` (also below) to better highlight this.

michalk8 commented on src/ott/solvers/nn/models.py:
```diff
@@ -0,0 +1,352 @@
+
+
+class NeuralTrainState(train_state.TrainState):
+  """Adds information about the model's value and gradient to the state."""
```
Currently, there's no documentation for this, see [here](https://ott-jax--219.org.readthedocs.build/en/219/solvers/_autosummary/ott.solvers.nn.models.NeuralTrainState.potential_value_fn.html#ott.solvers.nn.models.NeuralTrainState.potential_value_fn). Could you please add, e.g., 1 line for the docstring (maybe take from `ModelBase.potential_value_fn`).

bamos replied:
Added an "Args" section, although that page is still not showing anything. Using "Args" is at least consistent with [flax's docs](https://github.com/google/flax/blob/main/flax/training/train_state.py#L42) for `train_state`.

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -9,76 +9,131 @@
+  associated with the :math:`\alpha` measure with
+  an :class:`~ott.solvers.nn.models.ICNN`,
+  :class:`~ott.solvers.nn.models.MLP`,
+  or other
```
Pet peeve: don't like this almost empty line (same for the `where` below).

bamos replied:
Joined them (although I like having long paths on their own line since it makes quickly editing them slightly easier)

michalk8 commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -9,76 +9,131 @@
+  which is a combination further described in :cite:`amos:23`.
+
+  The :class:`~ott.solvers.nn.models.ModelBase` potentials for
+  ``neural_f`` and ``neural_g`` can 1) both provide the
```
Can you use rST list as:
```rst
can

    1. both provide the values of the ...
    2. one of them

The potential ...
```

bamos commented on src/ott/solvers/nn/neuraldual.py:
```diff
@@ -9,76 +9,131 @@
+  potential's value can be obtained via the Fenchel conjugate
+  as discussed in :cite:`amos:23`.
+  The potential's value or gradient mapping is specified via
+  the :meth:`~ott.solvers.nn.models.ModelBase.is_potential`
```
Thanks, I'm still getting used to these :)

bamos commented:
Something else interesting: the entropic potentials are looking really nice on this toy data! It could make a lot of sense to use the entropic potentials trained on a small set of samples to initialize the neural potentials. \cc @Apooladian

```Python
from ott.problems.nn import dataset
from ott.solvers.linear import sinkhorn
from ott.geometry.pointcloud import PointCloud
from ott.problems.linear.potentials import EntropicPotentials
from ott.problems.linear import linear_problem

train_dataloaders, valid_dataloaders, input_dim = dataset.create_gaussian_mixture_samplers(
    name_source="square_five", name_target="square_four")

eval_data_source = next(valid_dataloaders.source_iter)
eval_data_target = next(valid_dataloaders.target_iter)

geom = PointCloud(eval_data_source, eval_data_target, epsilon=0.001)
prob = linear_problem.LinearProblem(geom)
out = sinkhorn.solve(geom,
                     threshold=0.005, lse_mode=True, jit=False,
                     max_iterations=1000)
entpots = EntropicPotentials(out.f, out.g, prob)
entpots.plot_ot_map(eval_data_source, eval_data_target)
```

gives:

![image](https://user-images.githubusercontent.com/707462/218234953-15917172-51dd-470d-b016-e337fb15b1b1.png)
![image](https://user-images.githubusercontent.com/707462/218234902-077b9008-421b-4135-8354-11885f163721.png)

michalk8 commented on src/ott/problems/linear/potentials.py:
```diff
@@ -157,6 +161,144 @@ def tree_unflatten(
+        along each dimension
+      contourf_kwargs: additional kwargs passed into :meth:`~matplotlib.axes.Axes.contourf`
+
+    Returns: matplotlib figure and axis with the plots.
```
Can you put `matplotlib ...` on new line + indent? Think this is why it renders weird in the docs.

michalk8 commented on src/ott/problems/linear/potentials.py:
```diff
@@ -157,6 +161,144 @@ def tree_unflatten(
+      scatter_kwargs: additional kwargs passed into :meth:`~matplotlib.axes.Axes.scatter`
+      legend_kwargs: additional kwargs passed into :meth:`~matplotlib.axes.Axes.legend`
+
+    Returns: matplotlib figure and axis with the plots
```
Can you put `matplotlib ...` on new line + indent, like for the `Args:` section? Think this is why it renders weird in the docs.

michalk8 commented on docs/tutorials/notebooks/icnn_inits.ipynb:
```diff
@@ -300,7 +136,7 @@
    "metadata": {},
    "source": [
-    "Next, we define the architectures parameterizing the dual potentials $f$ and $g$. These need to be parameterized by ICNNs. You can adapt the size of the ICNNs by passing a sequence containing hidden layer sizes. While ICNNs are by default containing partially positive weights, we can solve the problem using approximations to this positivity constraint (via weight clipping and a weight penali ...
+    "Next, we define the architectures parameterizing the dual potentials $f$ and $g$. These need to be parameterized by ICNNs. You can adapt the size of the ICNNs by passing a sequence containing hidden layer sizes. While ICNNs are by default containing partially positive weights, we can solve the problem using approximations to this positivity constraint (via weight clipping and a weight penali ...
```
In `ICNN`, the argument is `pos_weights`, not `positive_weights`, can you please update the notebook?

michalk8 commented on docs/problems/nn.rst:
```diff
@@ -0,0 +1,15 @@
+    :toctree: _autosummary
+
+    dataset.create
+    dataset.Dataloaders
```
This is outdated, please update according to the name changes in `dataset.py`.

michalk8 commented on src/ott/solvers/nn/models.py:
```diff
@@ -0,0 +1,360 @@
+      :func:`jax.nn.initializers.normal`).
+    act_fn: choice of activation function used in network architecture
+      (needs to be convex, default: :obj:`jax.nn.relu`).
+    gaussian_map: data inputs of source and target measures for
```
Missing docstring for `pos_weights` (seems it was missing for a long time).

michalk8 commented on docs/solvers/nn.rst:
```diff
@@ -10,12 +10,23 @@ Neural Dual
 
-    icnn.ICNN
-    layers.PositiveDense
+    models.NeuralTrainState
```
I'd exclude `NeuralTrainState` from the docs, as it is too technical and doesn't provide much info (given it's well documented in `W2NeuralDual`).

michalk8 reviewed (changes requested):
Looks great, I got just a few minor comments (mostly regarding documentation). After these, will be happy to merge!

bamos commented:
> Looks great, I got just a few minor comments (mostly regarding documentation). After these, will be happy to merge!

Great! Just put them in at https://github.com/ott-jax/ott/pull/219/commits/45f33bedb25d7023c4b561f6982bd319395f0d43

michalk8 commented:
This is a very nice contribution @bamos , LGTM!
